### PR TITLE
Feature/imc word logging

### PR DESCRIPTION
**Pull request by martijnvandermarel**
Publishes IMC states to /march/imc_states, including statusword, detailed error, motion error and accompanying descriptions

Belongs to https://github.com/project-march/march-iv/pull/258 for the msg in march_shared_resources

### CODE REVIEW

File: march_hardware/src/IMotionCube.cpp [hunk 1]
```diff
@@ -208,191 +208,196 @@ void IMotionCube::setControlWord(uint16 controlWord)
   set_output_bit16(slaveIndex, this->mosiByteOffsets[IMCObjectName::ControlWord], controlwordu);
 }
 
-void IMotionCube::parseStatusWord(uint16 statusWord)
+std::string IMotionCube::parseStatusWord(uint16 statusWord)
```
Comment (Ishadijcks) [inline march_hardware/src/IMotionCube.cpp:219]: Do you want some whitespace between the strings? Newlines or at least spaces?

File: march_hardware/src/IMotionCube.cpp [hunk 2]
```diff
@@ -208,191 +208,196 @@ void IMotionCube::setControlWord(uint16 controlWord)
...
+    uint16 sixBitMask = 0b0000000001101111;
+
+    uint16 notReadyToSwitchOn = 0b0000000000000000;
+    uint16 switchOnDisabled = 0b0000000001000000;
+    uint16 readyToSwitchOn = 0b0000000000100001;
+    uint16 switchedOn = 0b0000000000100011;
+    uint16 operationEnabled = 0b0000000000100111;
+    uint16 quickStopActive = 0b0000000000000111;
+    uint16 faultReactionActive = 0b0000000000001111;
+    uint16 fault = 0b0000000000001000;
```
Comment (Ishadijcks) [inline march_hardware/src/IMotionCube.cpp]: Is this reviewed by someone?

File: march_hardware/src/IMotionCube.cpp [hunk 3]
```diff
@@ -208,191 +208,196 @@ void IMotionCube::setControlWord(uint16 controlWord)
...
+    }
+    else if (statusWordFiveBitMasked == faultReactionActive)
+    {
+        return "Fault Reaction Active";
+    }
+    else if (statusWordFiveBitMasked == fault)
+    {
+        return "Fault";
+    }
+    else return "Not in a recognized IMC state";
```
Comment (Ishadijcks) [inline march_hardware/src/IMotionCube.cpp]: Throw a ROS_WARN here?

File: march_hardware/src/Joint.cpp [hunk 4]
```diff
@@ -94,6 +96,27 @@ float Joint::getTemperature()
...
+    std::bitset<16> detailedErrorBits = this->iMotionCube.getDetailedError();
+    return_strings.push_back(detailedErrorBits.to_string());
+    std::bitset<16> motionErrorBits = this->iMotionCube.getMotionError();
+    return_strings.push_back(motionErrorBits.to_string());
+
+    return_strings.push_back(this->iMotionCube.getState(this->iMotionCube.getStatusWord()));
+    return_strings.push_back(this->iMotionCube.parseDetailedError(this->iMotionCube.getDetailedError()));
+    return_strings.push_back(this->iMotionCube.parseMotionError(this->iMotionCube.getMotionError()));
+
+    return return_strings;
```
Comment (Ishadijcks) [inline march_hardware/src/Joint.cpp]: Using a vector to return multiple values is discourages, consider creating your own object instead

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 5]
```diff
@@ -212,6 +227,14 @@ void MarchHardwareInterface::read(ros::Duration elapsed_time)
     joint_velocity_[i] = filters::exponentialSmoothing(joint_velocity, joint_velocity_[i], 0.2);
 
     ROS_DEBUG("Joint %s: read position %f", joint_names_[i].c_str(), joint_position_[i]);
+
+    this->updateIMotionCubeState(i);
+  }
+
+  // Publish IMotionCubeStates if possible
+  if (realtime_pubs_->trylock())
```
Comment (Ishadijcks) [inline march_hardware_interface/src/march_hardware_interface.cpp]: You only need to do `this->updateIMotionCubeState(i);` when you are actually planning on publishing

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 6]
```diff
@@ -316,4 +339,18 @@ void MarchHardwareInterface::updatePowerNet()
     }
   }
 }
+
+void MarchHardwareInterface::updateIMotionCubeState(int jointIndex)
```
Comment (Ishadijcks) [inline march_hardware_interface/src/march_hardware_interface.cpp]: I suggest wrapping this in another method that updates it for all joints for clarity

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 7]
```diff
@@ -316,4 +339,18 @@ void MarchHardwareInterface::updatePowerNet()
...
+{
+    std::vector<std::string> IMotionCubeState = marchRobot.getJoint(joint_names_[jointIndex]).getIMotionCubeState();
+
+    realtime_pubs_->msg_.joint_names.push_back(joint_names_[jointIndex]);
+    realtime_pubs_->msg_.status_word.push_back(IMotionCubeState[0]);
+    realtime_pubs_->msg_.detailed_error.push_back(IMotionCubeState[1]);
+    realtime_pubs_->msg_.motion_error.push_back(IMotionCubeState[2]);
+    realtime_pubs_->msg_.state.push_back(IMotionCubeState[3]);
+    realtime_pubs_->msg_.detailed_error_description.push_back(IMotionCubeState[4]);
+    realtime_pubs_->msg_.motion_error_description.push_back(IMotionCubeState[5]);
```
Comment (Ishadijcks) [inline march_hardware_interface/src/march_hardware_interface.cpp]: Here you see why a custom object is desirable. There is nothing that guarantees this order is correct

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 8]
```diff
@@ -316,4 +339,18 @@ void MarchHardwareInterface::updatePowerNet()
     }
   }
 }
+
+void MarchHardwareInterface::updateIMotionCubeState(int jointIndex)
+{
+    std::vector<std::string> IMotionCubeState = marchRobot.getJoint(joint_names_[jointIndex]).getIMotionCubeState();
```
Comment (Ishadijcks) [inline march_hardware_interface/src/march_hardware_interface.cpp]: Do not start variables with a capital letter

**Review by Ishadijcks — CHANGES_REQUESTED** [review]

File: march_hardware/src/IMotionCube.cpp (hunk 3 above)
Comment (martijnvandermarel) [reply march_hardware/src/IMotionCube.cpp]: I've seen this happen before going to operation enabled, so will probably always be thrown before startup

File: march_hardware/src/IMotionCube.cpp (hunk 2 above)
Comment (martijnvandermarel) [reply march_hardware/src/IMotionCube.cpp]: Bart has seen this, but not too thoroughly I think

File: march_hardware/include/march_hardware/Joint.h [hunk 9]
```diff
@@ -3,13 +3,28 @@
...
+#include <vector>
 
 #include <march_hardware/IMotionCube.h>
 #include <march_hardware/PowerDistributionBoard.h>
 #include <march_hardware/TemperatureGES.h>
 
 namespace march4cpp
 {
+
+class IMotionCubeState
```
Comment (Ishadijcks) [inline march_hardware/include/march_hardware/Joint.h]: Move to own class

**Review by Ishadijcks — APPROVED** [review]

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 10]
```diff
@@ -4,18 +4,20 @@
 #include <joint_limits_interface/joint_limits_interface.h>
 #include <joint_limits_interface/joint_limits_urdf.h>
 #include <sstream>
+#include <bitset>
```
Comment (Ishadijcks) [inline march_hardware_interface/src/march_hardware_interface.cpp]:
```suggestion
```

**Pull request merged**